dependent_types.md

```diff
@@ -0,0 +1,386 @@
+
+
+**Types depending on Types** (_type functions_)
+
+In order to demonstrate what we mean by this we will use Haskell extension called Type Families. 
+
+Type Families provide a way to define multiple type constructors with different types in place of a type parameter.
+Here is an example from Haskell [wiki](https://wiki.haskell.org/Haskell)
+
+```
+-- Declare a list-like data family
+data family XList a
+ 
+-- Declare a list-like instance for Char
+data instance XList Char = XCons !Char !(XList Char) | XNil
+ 
+-- Declare a number-like instance for ()
+data instance XList () = XListUnit !Int
+```
+
+Here we are providing two instances of the same data type. One can be constructed when using `Char` for the type constructor parameter and the other one when using  `()`. Using Type Families we can say that type used as a parameter to a type constructor determines the type we will be able to construct - so type depending on another type. 
+
+Haskell is famous for its _If it compiles - it works_ approach, which is very true but we are haskellers - we always want more type safety and abstraction. 
+
+This leads us to the final step to dependent types which is: 
+
+**Types depending on Values** (_dependent types_)
+
+As I mentioned at the beginning Haskell still does not have native support for dependent types but we have a handful of extensions that provide a way to touch on that. We are now arriving to the next step in understanding dependent types which is
+
+## Local Assumptions
+
+GADTs or Generalized Algebraic DataTypes is the Haskell extension that provides us with _local assumptions_. What does that mean? 
+
+GADTs provide us with a way to have richer return types than vanilla ADTs. If we look at the GHC manual example
+
+```
+data Term a where
+    Lit    :: Int -> Term Int
+    Succ   :: Term Int -> Term Int
+    IsZero :: Term Int -> Term Bool
+    If     :: Term Bool -> Term a -> Term a -> Term a
+    Pair   :: Term a -> Term b -> Term (a,b)
+```
+We notice that the return type is not always `Term a` as it would be with normal ADTs. When pattern matching on type constructors that are constructed using GADTs syntax we are able to have type refinement to specific constructor so `a` can be refined to `Int` in following function
+
+```
+eval :: Term a -> a
+eval Lit a = ... 
+```
+Since pattern matching reduced to this branch and return type is `Term Int` 
+
+```
+Lit    :: Int -> Term Int
+```
+
+When pattern matching on type constructor using GADTs we get assumption that the type was constructed with certain type parameter/s. This is what we also get with data families but difference is that we get _local_ assumptions since we are able to use pattern matching on polymorphic type variables , so all possible results defined in GADt since they have fixed set of constructors.
```

Comment:
Oh, now I see why you used a `data family` as an example above — you now compare data families to GADTs. Okay, in this case consider just a second example there, with a type family used as a type level function on type level data. For instance, `type family Reverse (xs :: [a]) :: [a]` to reverse a list.